File: restclient-python/src/RestClientRentals.py
import requests

import base64


class RestClientRentals:
    
   
    def __init__(self):
        ''' 
        Client ID and secret that identifies this client
        for the REST API
        '''
        self.clientId = 'rentals-client-id'
        self.clientSecret = 'ddkl44#SDsfg'
    
        self.urlRoot = 'http://localhost:8080'
    
    
    def getAccessToken(self, email):
        credentials = self.clientId + ':' + self.clientSecret
        headers = {
            'Authorization': 'Basic ' + base64.b64encode(credentials),
            'Content-Type': 'application/x-www-form-urlencoded'}
        url = self.urlRoot + '/oauth/token'
        data = {
            'grant_type': 'password',
            'username': email,
            'password': 'pass',
            'scope': 'read write'}
        request = requests.Request('POST', url, data = data, headers = headers)
        prepRequest = request.prepare()
        
        RestClientRentals.pretty_print_POST(prepRequest)
        
        session = requests.Session()
        response = session.send(prepRequest)
       
        return response.json()['access_token']
        
    @staticmethod
    def addHeaderForAccessToken(headers, accessToken):
        headers['Authorization'] = 'Bearer ' + accessToken
    
    '''
    Makes a POST request to search for all users with an email 
    address @rentals.io
    '''
    def post_searchUsers(self, accessToken):
            
        url = self.urlRoot + '/api/rentals/users/search'
        data = '''{
            "email": "@rentals.io"
        }'''
        headers = {
            'Content-Type': 'application/json'
        }
        RestClientRentals.addHeaderForAccessToken(headers, accessToken)
        request = requests.Request('POST', url, data = data, headers = headers)
        prepRequest = request.prepare()
        
        RestClientRentals.pretty_print_POST(prepRequest)
        
        session = requests.Session()
        response = session.send(prepRequest)
        
        print(str(response.content))
    
    
    '''
    Makes a POST request to add a user
    '''
    def post_addUser(self, accessToken):
            
        url = self.urlRoot + '/api/rentals/users/'
        data = '''{
                    "email":"test1@rentals.com", 
                    "password":"pass", 
                    "roles": [{
                         "id": 2, 
                         "name":"CLIENT"
                     }]
                }'''
        headers = {
            'Content-Type': 'application/json'
        }
        RestClientRentals.addHeaderForAccessToken(headers, accessToken)
        request = requests.Request('POST', url, data = data, headers = headers)
        prepRequest = request.prepare()
        
        RestClientRentals.pretty_print_POST(prepRequest)
        
        session = requests.Session()
        response = session.send(prepRequest)
        
        if response.status_code != 201:
            raise Exception('Failed to add user. Status code: {}'.format(response.status_code))
        
        ''' extract the user id from the Location URL '''
        newUserUrl = response.headers['Location']
        lst = newUserUrl.split('/')
        return lst[len(lst) - 1]
    
    
    '''
    Makes a POST request to add a user
    '''
    def put_updateUser(self, accessToken, userId):
            
        url = self.urlRoot + '/api/rentals/users/' + userId
        data = '''{
                    "email":"test1@rentals.com", 
                    "password":"pass1", 
                    "roles": [
                        {
                         "id": 2, 
                         "name":"CLIENT"
                         },
                         {
                          "id": 1, 
                          "name":"CLERK"
                         }
                    ]
                }'''
        headers = {
            'Content-Type': 'application/json'
        }
        RestClientRentals.addHeaderForAccessToken(headers, accessToken)
        request = requests.Request('PUT', url, data = data, headers = headers)
        prepRequest = request.prepare()
        
        RestClientRentals.pretty_print_POST(prepRequest)
        
        session = requests.Session()
        response = session.send(prepRequest)
        
        if response.status_code != 204:
            raise Exception('Failed to update user. Status code: {}'.format(response.status_code))
        
        
    '''
    Makes a DELETE request to delete a user
    '''
    def delete_deleteUser(self, accessToken, userId):
            
        url = self.urlRoot + '/api/rentals/users/' + userId
        headers = {
            'Content-Type': 'application/json'
        }
        RestClientRentals.addHeaderForAccessToken(headers, accessToken)
        request = requests.Request('DELETE', url, headers = headers)
        prepRequest = request.prepare()
        
        RestClientRentals.pretty_print_POST(prepRequest)
        
        session = requests.Session()
        response = session.send(prepRequest)
        
        if response.status_code != 204:
            raise Exception('Failed to update user. Status code: {}'.format(response.status_code))
        
    '''
    Makes a GET request to retrieve a user
    '''
    def get_getUser(self, accessToken, userId):
            
        url = self.urlRoot + '/api/rentals/users/' + userId
        headers = {
            'Content-Type': 'application/json'
        }
        RestClientRentals.addHeaderForAccessToken(headers, accessToken)
        request = requests.Request('GET', url, headers = headers)
        prepRequest = request.prepare()
        
        RestClientRentals.pretty_print_POST(prepRequest)
        
        session = requests.Session()
        response = session.send(prepRequest)
        
        if response.status_code != 200:
            raise Exception('Failed to update user. Status code: {}'.format(response.status_code))
        
    
    
    
    '''
    https://stackoverflow.com/questions/20658572/python-requests-print-entire-http-request-raw
    '''
    @staticmethod
    def pretty_print_POST(req):
    
        print('{}\n{}\n{}\n\n{}'.format(
            '-----------START-----------',
            req.method + ' ' + req.url,
            '\n'.join('{}: {}'.format(k, v) for k, v in req.headers.items()),
            req.body,
        ))
        
    '''
    Runs all methods covered in this post
    '''
    def runAll(self):
        accessToken = self.getAccessToken('admin@rentals.io')
        self.post_searchUsers(accessToken)
        newUserId = self.post_addUser(accessToken)
        self.get_getUser(accessToken, newUserId)
        self.put_updateUser(accessToken, newUserId)
        self.delete_deleteUser(accessToken, newUserId)


inst = RestClientRentals()
inst.runAll()

File: restclient-js-nodejs/RestClientRentals.js

var req = require('request');

var clientId = 'rentals-client-id';
var clientSecret = 'ddkl44#SDsfg';

var credentials = clientId + ':' + clientSecret;

/*
 * Retrieves an OAuth2 access token given a user's email
 */
function getAccessToken(email, callback) {
	
	const options =  {
			method: 'POST',
		    uri: 'http://localhost:8080/oauth/token',
		    form: {
		    	grant_type: 'password',
			    username: email,
			    password: 'pass',
			    scope: 'read write'
		    },
		    headers: {
		    	'Authorization': 'Basic ' + Buffer.from(credentials).toString('base64'),
	    		'Content-Type': 'application/x-www-form-urlencoded'
		    },
		    // parse JSON and convert into JS object
		    json: true
		};
	
	req(options, (error, response, body) => {
		if(error) {
			callback(error)
		} else if(response.statusCode != 200) {
			callback(new Error('Response status is ' + response.statusCode + '. Body is ' 
					+ JSON.stringify(body)));
		} else {
			callback(error, body.access_token)
		}
	})
}

/*
 * Makes a POST request to search for all users with an email 
 * address @rentals.io
 */
function post_searchUsers(token, callback) {
	const options =  {
			method: 'POST',
		    uri: 'http://localhost:8080/api/rentals/users/search',
		    body: {
		    	email: '@rentals.io'
		    },
		    headers: {
		    	'Authorization': 'Bearer ' + token,
	    		'Content-Type': 'application/json'
		    },
		    // parse JSON and convert into JS object
		    json: true
		};
	
	req(options, (error, response, body) => {
		if(error) {
			callback(error)
		} else if(response.statusCode != 200) {
			callback(new Error('Response status is ' + response.statusCode + '. Body is ' 
					+ JSON.stringify(body)));
		} else {
			callback(error, body)
		}
	})
}

/*
 * Makes a POST request to add a user
 */
function post_addUser(token, callback) {
	const options =  {
			method: 'POST',
		    uri: 'http://localhost:8080/api/rentals/users',
		    body: {
		    	email:"test1@rentals.com", 
                password:"pass", 
                roles: [{
                     id: 2, 
                     name:"CLIENT"
                 }]
		    },
		    headers: {
		    	'Authorization': 'Bearer ' + token,
	    		'Content-Type': 'application/json'
		    },
		    // parse JSON and convert into JS object
		    json: true
		};
	
	req(options, (error, response, body) => {
		if(error) {
			callback(error)
		} else if(response.statusCode != 201) {
			callback(new Error('Response status is ' + response.statusCode + '. Body is ' 
					+ JSON.stringify(body)));
		} else {
			// extract userId from the Location header
			var locArr = response.headers.location.split('/');		
			callback(error, locArr[locArr.length - 1]);
		}
	})
}

/*
 * Makes a PUT request to update a user
 */
function put_updateUser(token, userId, callback) {
	const options =  {
			method: 'PUT',
		    uri: 'http://localhost:8080/api/rentals/users/' + userId,
		    body: {
		    	email:"test1@rentals.com", 
                password:"pass1", 
                roles: [{
                     id: 2, 
                     name:"CLIENT"
                 },
                 {
                     id: 1, 
                     name:"CLERK"
                 }]
		    },
		    headers: {
		    	'Authorization': 'Bearer ' + token,
	    		'Content-Type': 'application/json'
		    },
		    // parse JSON and convert into JS object
		    json: true
		};
	
	req(options, (error, response, body) => {
		if(error) {
			callback(error)
		} else if(response.statusCode != 204) {
			callback(new Error('Response status is ' + response.statusCode + '. Body is ' 
					+ JSON.stringify(body)));
		} else {		
			callback(error);
		}
	})
}

/*
 * Makes a DELETE request to delete a user
 */
function delete_deleteUser(token, userId, callback) {
	const options =  {
			method: 'DELETE',
		    uri: 'http://localhost:8080/api/rentals/users/' + userId,
		    headers: {
		    	'Authorization': 'Bearer ' + token,
		    	'Content-Type': 'application/json'
		    },
		    // parse JSON and convert into JS object
		    json: true
		};
	
	req(options, (error, response, body) => {
		if(error) {
			callback(error)
		} else if(response.statusCode != 204) {
			callback(new Error('Response status is ' + response.statusCode + '. Body is ' 
					+ JSON.stringify(body)));
		} else {		
			callback(error);
		}
	})
}

/*
 * Makes a GET request to retrieve a user
 */
function get_getUser(token, userId, callback) {
	const options =  {
			method: 'GET',
		    uri: 'http://localhost:8080/api/rentals/users/' + userId,
		    headers: {
		    	'Authorization': 'Bearer ' + token,
		    	'Content-Type': 'application/json'
		    },
		    // parse JSON and convert into JS object
		    json: true
		};
	
	req(options, (error, response, body) => {
		if(error) {
			callback(error)
		} else if(response.statusCode != 200) {
			callback(new Error('Response status is ' + response.statusCode + '. Body is ' 
					+ JSON.stringify(body)));
		} else {		
			callback(error, body);
		}
	})
}

/*
 * Run all methods covered in the post.
 */
function runAll(token) {
	// search for users first
	post_searchUsers(token, (error, body) => {
		if(error) {
			console.log(error);
		} else {
			console.log(body);
			
			// add a user now
			post_addUser(token, (error, userId) => {
				if(error) {
					console.log(error);
				} else {
					console.log(userId);
					
					// update the password for the newly added user
					put_updateUser(token, userId, (error) => {
						if(error) {
							console.log(error);
						} else {
							console.log('User ' + userId + ' updated successfuly')
							
							// get the user now
							get_getUser(token, userId, (error, body) => {
								if(error) {
									console.log(error);
								} else {
									console.log('User : ' + userId + ': ' + JSON.stringify(body));
									
									// delete the user now
									delete_deleteUser(token, userId, (error) => {
										if(error) {
											console.log(error);
										} else {
											console.log('User ' + userId + ' deleted successfuly');
										}
									})
								}
							})
						}
					})
				}
			})
		}
	});
	
}

/*
 * Get an access token first and use it for all methods
 */
getAccessToken('admin@rentals.io', (err, token) => {
	if(err) {
		console.log(err);
	} else {
		runAll(token);
	}
});

